src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,961 @@
+                                            level, tile_col, tile_row,
+                                            &cache_entry);
+  if (!tiledata) {
+    g_auto(_openslide_slice) box =
+      _openslide_slice_alloc(tile_size * tile_size * 4);
+    guint32 frame_number = 1 + tile_col + l->tiles_across * tile_row;
+    if (frame_number < 1 || frame_number > l->num_frames) {
+      g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                  "Frame number out of range %d - %d",
+                  1,
+                  l->num_frames);
+      return false;
+    }
+
+    g_mutex_lock(&l->file->lock);
+
+    DcmError *dcm_error = NULL;
+    DcmFrame *frame = dcm_filehandle_read_frame(&dcm_error,
+                                                l->file->filehandle, 
+                                                l->file->metadata,
+                                                l->file->bot, 
+                                                frame_number);
+
+    g_mutex_unlock(&l->file->lock);
+
+    if (frame == NULL) {
+      set_gerror_from_dcm_error(err, &dcm_error);
+      return false;
+    }
+
+    const char *frame_value = dcm_frame_get_value(frame);
+    uint32_t frame_length = dcm_frame_get_length(frame);
+
+#ifdef DEBUG
+    print_frame(frame);
+#endif /*DEBUG*/
+
+    if (!_openslide_jpeg_decode_buffer(frame_value,
+                                       frame_length,
+                                       box.p,
+                                       dcm_frame_get_columns(frame),
+                                       dcm_frame_get_rows(frame),
+                                       err)) {
+      return false;
+    }
+
+    // clip, if necessary
+    if (!_openslide_clip_tile(box.p,
+                              tile_size, tile_size,
+                              l->base.w - tile_col * tile_size,
+                              l->base.h - tile_row * tile_size,
+                              err)) {
+      return false;
+    }
+
+    // put it in the cache
+    tiledata = _openslide_slice_steal(&box);
+    _openslide_cache_put(osr->cache,
+			 level, tile_col, tile_row,
+			 tiledata, tile_size * tile_size * 4,
+			 &cache_entry);
+  }
+
+  // draw it
+  g_autoptr(cairo_surface_t) surface =
+    cairo_image_surface_create_for_data((unsigned char *) tiledata,
+                                        CAIRO_FORMAT_ARGB32,
+                                        tile_size, tile_size, tile_size * 4);
+  cairo_set_source_surface(cr, surface, 0, 0);
+  cairo_paint(cr);
+
+  return true;
+}
+
+static bool paint_region(openslide_t *osr, 
+                         cairo_t *cr,
+                         int64_t x, 
+                         int64_t y,
+                         struct _openslide_level *level,
+                         int32_t w, 
+                         int32_t h,
+                         GError **err) {
+  struct dicom_ops_data *data = osr->data;
+  Level *l = (Level *) level;
```

Comment:
Please use typedefs in accordance with the [style guide](https://openslide.org/docs/newformat/#memory-allocation) (i.e., minimally).

Reply:
OK, done (I hope).